services/puppetforge/puppetforge-base.js

```diff
@@ -24,6 +24,8 @@ const modulesSchema = Joi.object({
   ),
 }).required()
 
+const modulesValidationSchema = Joi.array().required()
```

Comment:
This is an interesting API response format we're working with here.
I think the current approach to dealing with it is a bit brittle in a couple of ways:
1. We're not really validating the input properly. I accept this is difficult - Joi doesn't make it super easy to validate an array containing objects which all match different schemas.
2. We're relying on the position of the object in the array.
I think my suggestion would be to define the schema like this:

```js
import { nonNegativeInteger } from '../validators.js'

const schema = Joi.array()
  .items(
    Joi.alternatives().try(
      Joi.object({
        name: Joi.string().valid('total').required(),
        score: nonNegativeInteger,
      }).required(),
      Joi.object({}).required(),
    ),
  )
  .custom((value, helpers) => {
    // Custom validation to check for exactly one type1 object
    const totalCount = value.filter(item => item.name === 'total').length
    if (totalCount !== 1) {
      return helpers.message(
        'Array must contain exactly one object of type "total"',
      )
    }
    return value
  })
```

This requires the array to have exactly one object in it with a `name` field equal to `"total"` and a `score field`.

Then instead of relying on position within the array, we locate the relevant object using:

```js
data.find(el => el.name == 'total')
```

and we've already validated that will exist with the schema.

Reply:
Nice!